### DISCUSSION TITLE
force language detection for files in .gitignore

### DISCUSSION THREAD

**Discussion by Icyrockton** [github-linguist/linguist#6611] [Q&A] [answered]
I'm implementing a crawler.
I want to use linguistic command line single-file mode to detect the language of certain downloaded files, but the directory of these downloaded files are annotated in `.gitignore` (e.g., download/ ), and linguistic will not be able to find these files and report an error, but locally I own these files, can I force a bypass of the `.gitignore` restriction?

```
Traceback (most recent call last):
        8: from /usr/local/bin/github-linguist:23:in `<main>'
        7: from /usr/local/bin/github-linguist:23:in `load'
        6: from /var/lib/gems/2.7.0/gems/github-linguist-7.27.0/bin/github-linguist:152:in `<top (required)>'
        5: from /var/lib/gems/2.7.0/gems/github-linguist-7.27.0/bin/github-linguist:109:in `github_linguist'
        4: from /var/lib/gems/2.7.0/gems/github-linguist-7.27.0/lib/linguist/blob_helper.rb:155:in `text?'
        3: from /var/lib/gems/2.7.0/gems/github-linguist-7.27.0/lib/linguist/blob_helper.rb:127:in `binary?'
        2: from /var/lib/gems/2.7.0/gems/github-linguist-7.27.0/lib/linguist/lazy_blob.rb:83:in `data'
        1: from /var/lib/gems/2.7.0/gems/github-linguist-7.27.0/lib/linguist/lazy_blob.rb:109:in `load_blob!'
/var/lib/gems/2.7.0/gems/github-linguist-7.27.0/lib/linguist/lazy_blob.rb:109:in `to_buffer': wrong argument type nil (expected String) (TypeError)
```

a simple reproduce repo https://github.com/Icyrockton/linguistic_test

```python
python main.py
```

**Comment by lildude** [comment, marked as answer]
Your repo link is inaccessible.

> can I force a bypass of the .gitignore restriction?

No. Linguist expects to work on files in a git repo by default and will always look first to see if the file is in a repo. If the files are in a repo but in a directory ignored by git, then Linguist (technically [rugged](https://github.com/libgit2/rugged) which does the repo accessing etc) won't be able to find the files and you get the behaviour you are seeing. 

This is the section of code doing this (lines 97-104):

https://github.com/github-linguist/linguist/blob/7ca3799b8b5f1acde1dd7a8dfb7ae849d3dfb4cd/bin/github-linguist#L97-L107

The only way around this is to put the files in a directory completely outside of any git repo which will then be taken care of by line 106 above.

**Reply by Icyrockton** [reply]
My mistake!  I have changed the visibility to public.
Since the downloaded files are at the 100GB level, redundant copy is very trivial, I just wanted to use lingstuic's file language detection feature, which is the best library I've found to use.
Thanks for your reply by the way!  :)

**Reply by lildude** [reply]
If you're not tied to using Ruby, you might be interested in [go-enry](https://github.com/go-enry/go-enry). This is written in Go and feeds directly off Linguist and is what GitHub's Search functionality uses. I'm not sure if it has the same repo requirements and limitations, but it's probably worth a look.

**Reply by Icyrockton** [reply]
Thank you for your help, I will check it out ❤️
